MF-1425 - Enhancement for external UUID

aspnair opened the pull request:
MF-1425 Enhancement to supply an external UUID for Things and Channels.
- This entire feature would be switched off by default, and enabled by the
  new environment MF_THINGS_ALLOW_EXT_ID.
- This environment is passed to the Things Service layer.
- The IDProvider interface is updated to include FromString() to validate
  the externally supplied UUID.
- A new local function createUid() is added to either generate a fresh UUID
  or to extract and validate the supplied UUID in the Things Service layer.
- While creating a Thing or Channel if an "extid" tag is supplied inside "metadata",
  then the supplied external UUID is taken after validation.
- If this "extid" is not supplied the default method of UUID generation is used.

Usage:
$ uuidgen
99336f21-bd45-4e47-87ab-f57505bf5ef2

$ curl -s -S -i -X POST -H "Content-Type: application/json" -H "Authorization: ${TOKEN}" http://localhost/things -d '{"name": "u1_t1", "metadata": {"extid": "99336f21-bd45-4e47-87ab-f57505bf5ef2", "type": "device", "desc": "user1 thing1"}}'

$ curl -s -S -i -X GET -H "Content-Type: application/json" -H "Authorization: ${TOKEN}" http://localhost/things
{"total":1,"offset":0,"limit":10,"order":"","direction":"","things":[{"id":"99336f21-bd45-4e47-87ab-f57505bf5ef2","name":"u1_t1","key":"a87563c2-7a35-449d-b50a-c7a261a61aa5","metadata":{"desc":"user1 thing1","extid":"99336f21-bd45-4e47-87ab-f57505bf5ef2","type":"device"}}]}

Pull request title should be `MF-XXX - description` or `NOISSUE - description` where XXX is ID of issue that this PR relate to.
Please review the [CONTRIBUTING.md](./CONTRIBUTING.md) file for detailed contributing guidelines.

### What does this do?

### Which issue(s) does this PR fix/relate to?
Put here `Resolves #XXX` to auto-close the issue that your PR fixes (if such)

### List any changes that modify/break current functionality

### Have you included tests for your changes?

### Did you document any new/modified functionality?

### Notes

manuio commented on pkg/uuid/mock.go:
```diff
@@ -28,6 +29,11 @@ func (up *uuidProviderMock) ID() (string, error) {
 }
 
+func (up *uuidProviderMock) FromString(input string) (string, error) {
+
```
kill this blank line

aspnair replied:
done

manuio commented on pkg/uuid/mock.go:
```diff
@@ -28,6 +29,11 @@ func (up *uuidProviderMock) ID() (string, error) {
 
+func (up *uuidProviderMock) FromString(input string) (string, error) {
+
+	return "", errors.New("Error uuidProviderMock does not implement FromString")
```
Do we need this error? Maybe simply `nil` here

aspnair replied:
I also thought of removing it. But interesting when I changed it into
return "", nil
the go test was failing.  Let me try more and will update you on that.  For now, I left it as it is.

manuio commented on things/service.go:
```diff
@@ -147,6 +147,22 @@ type thingsService struct {
 
+// createUID: take the external id if supplied, otherwise use default provider
+// validate the external id, if validation fails, generate an id using the
+// default provider.
```
You can probably simply this to something like
```
provideUUID: Generates UUID or validates external ID if supplied.

aspnair replied:
done

manuio commented on things/service.go:
```diff
@@ -147,6 +147,22 @@ type thingsService struct {
+// validate the external id, if validation fails, generate an id using the
+// default provider.
+func (ts *thingsService) createUID(extid string) (id string, err error) {
+	if(len(extid) > 0) {
```
Avoid if/else and don't ignore errors. Also, I don't think that is good to generate an UUID if exterID is wrong. Better return an error. Something like: 
```
	if extID != "" {
		id, err = ts.idProvider.FromString(extID)
		if err != nil {
			return "", err
		}
		return id, nil
	}

	id, err = ts.idProvider.ID()
	if err != nil {
		return "", err
	}

	return id, nil
```

In any case, check my previous comment please. You can probably remove this function and simply validate the format in `request.go`

aspnair replied:
I have a few comments about it, will put in  a separate comment in some time. :-)

manuio commented on things/service.go:
```diff
@@ -147,6 +147,22 @@ type thingsService struct {
+// createUID: take the external id if supplied, otherwise use default provider
+// validate the external id, if validation fails, generate an id using the
+// default provider.
+func (ts *thingsService) createUID(extid string) (id string, err error) {
```
Maybe better `provideUUID` and `externalID` or `extID`

aspnair replied:
changed to extID. Thank You.

manuio commented on pkg/sdk/go/things_test.go:
```diff
@@ -85,14 +85,14 @@ func TestCreateThing(t *testing.T) {
 			token:    token,
 			err:      nil,
-			location: "001",
+			location: "10000001-114b-4924-968e-239eb242cef1",
```
Use `thing.ID` here

manuio commented on pkg/sdk/go/things_test.go:
```diff
@@ -85,14 +85,14 @@ func TestCreateThing(t *testing.T) {
 			token:    token,
 			err:      nil,
-			location: "002",
+			location: "123e4567-e89b-12d3-a456-000000000002",
```
Here you should use the ID returned on creation

manuio commented on things/api/things/http/endpoint.go:
```diff
@@ -22,6 +22,7 @@ func createThingEndpoint(svc things.Service) endpoint.Endpoint {
 
 		th := things.Thing{
 			Key:      req.Key,
+			ID:       req.ID,
```
Create a new object here `th2` with `ID`  and remove the `ID` from `th`

manuio commented on things/api/things/http/endpoint.go:
```diff
@@ -248,7 +250,7 @@ func createChannelEndpoint(svc things.Service) endpoint.Endpoint {
 		}
 
-		ch := things.Channel{Name: req.Name, Metadata: req.Metadata}
+		ch := things.Channel{Name: req.Name, ID: req.ID, Metadata: req.Metadata}
```
Create a new test case instead of mofying previous ones. ID shouldn't be mandatory

manuio commented on things/service.go:
```diff
@@ -147,6 +147,30 @@ type thingsService struct {
+	}
+
+	return id, err
+*/
```
Remove dead code

aspnair replied:
That was my mistake.  Let me remove that immediately.

manuio commented on things/api/things/http/requests.go:
```diff
@@ -21,14 +22,33 @@ type createThingReq struct {
 
+// validateUUID: validates the externally supplied UUID as per standard
+func validateUUID(extID string) (err error) {
+	if(extID == "") {
```
`if extID == "" {`

aspnair replied:
done

manuio commented on things/api/things/http/requests.go:
```diff
@@ -21,14 +22,33 @@ type createThingReq struct {
+	}
+
+	id, err := uuid.FromString(extID)
+	if((id.String() != extID) || (err != nil)) {
```
`if id.String() != extID || err != nil`

aspnair replied:
done

manuio commented on things/api/things/http/requests.go:
```diff
@@ -21,14 +22,33 @@ type createThingReq struct {
 		return things.ErrUnauthorizedAccess
 	}
 
+	if(validateUUID(req.ID) != nil) {
```
`if validateUUID(req.ID) != nil`

aspnair replied:
done

manuio commented on things/api/things/http/requests.go:
```diff
@@ -51,6 +71,10 @@ func (req createThingsReq) validate() error {
 	}
 
 	for _, thing := range req.Things {
+		if(validateUUID(thing.ID) != nil) {
```
`if validateUUID(thing.ID) != nil`

aspnair replied:
done

manuio commented on things/api/things/http/requests.go:
```diff
@@ -133,6 +162,10 @@ func (req createChannelsReq) validate() error {
 	}
 
 	for _, channel := range req.Channels {
+		if(validateUUID(channel.ID) != nil) {
```
`if validateUUID(channel.ID) != nil`

aspnair replied:
done

manuio commented on things/service.go:
```diff
@@ -147,6 +147,17 @@ type thingsService struct {
 
+// provideUUID: Generates UUID or takes external ID if supplied.
+func (ts *thingsService) provideUUID(extID string) (id string, err error) {
+	if(extID == "") {
```
`if extID == ""`

aspnair replied:
done

manuio commented on pkg/sdk/go/channels_test.go:
```diff
@@ -67,6 +71,20 @@ func TestCreateChannel(t *testing.T) {
+		},
+		{
+			desc:   "create a new channel with wrong external UUID",
+			channel: chWithWrongExtID,
```
Maybe you can short this name to `chWrongExtID`

aspnair replied:
Done

manuio commented on pkg/sdk/go/channels_test.go:
```diff
@@ -67,6 +71,20 @@ func TestCreateChannel(t *testing.T) {
 		},
+		{
+			desc:   "create a new channel with external UUID",
+			channel: channel1,
```
To be consistent with other naming this should be `chExtID`

aspnair replied:
Done

manuio commented on pkg/sdk/go/things_test.go:
```diff
@@ -166,6 +175,20 @@ func TestCreateThings(t *testing.T) {
 		},
+		{
+			desc:   "create new things with external UUID",
+			things: thingsWithExtID,
```
I would short this name to `thExtID`

aspnair replied:
Done

manuio commented on pkg/sdk/go/things_test.go:
```diff
@@ -166,6 +175,20 @@ func TestCreateThings(t *testing.T) {
+		},
+		{
+			desc:   "create new things with wrong external UUID",
+			things: thingsWithWrongExtID,
```
Same here `thWrongExtID`

aspnair replied:
Done

manuio commented on scripts/ci.sh:
```diff
@@ -82,8 +83,13 @@ setup_lint() {
+        setup_protoc
+    else
+        echo "Ignoring Go and Protoc setup"
+    fi
```
Can you explain us why we need this pelase?

aspnair replied:
Will write a detailed explanation in the comment below separately.

manuio commented on pkg/sdk/go/channels_test.go:
```diff
@@ -41,21 +45,21 @@ func TestCreateChannel(t *testing.T) {
 		{
 			desc:    "create new channel",
-			channel: channel,
+			channel: channel1,
```
Here you should not use a Channel with external ID. Create a new one `chExtID` and use it only in the corresponding tests.

aspnair replied:
Done

manuio commented on things/service_test.go:
```diff
@@ -43,8 +44,18 @@ func newService(tokens map[string]string) things.Service {
 func TestCreateThings(t *testing.T) {
 	svc := newService(map[string]string{token: email})
+	meta := things.Metadata{}
+	meta["type"] = "extid"
```
Not sure if this is needed

aspnair replied:
Removed Meta.

manuio commented on things/service_test.go:
```diff
@@ -24,10 +24,11 @@ const (
 
 var (
-	thing   = things.Thing{Name: "test"}
+	thingList = [n]things.Thing{}
```
Here you can do `things = []things.Thing{{Name: "thing1"},{Name: "thin2"}}`
No need to use entities with external ID for simple Thing creation.

aspnair replied:
The reason is that, a new function has been included by me TestInit() in the original patch.  We have to have Name, ID and Key, that too for 10 Things.  I thought it would be too long to keep all these 10 things with associated Name, ID and Key in the var list.  If required, we can also do the way you suggested.

manuio commented on things/service_test.go:
```diff
@@ -64,6 +75,20 @@ func TestCreateThings(t *testing.T) {
 		},
+		{
+			desc:   "create new things with external UUID",
+			things: []things.Thing{{ID: "fe6b4e92-cc98-425e-b0aa-000000000001", Name: "a", Metadata: meta},
```
Create a `var` for this lists. For example `thingsExtID`

aspnair replied:
User var.

manuio commented on things/service_test.go:
```diff
@@ -557,6 +582,20 @@ func TestCreateChannels(t *testing.T) {
 		},
+		{
+			desc:   "create new channels with external UUID",
+			channels: []things.Channel{{ID: "fe6b4e92-cc98-425e-b0aa-000000000001", Name: "a", Metadata: meta},
```
Same here, create a `var` for these lists

aspnair replied:
Used var here too for channel

manuio commented on pkg/sdk/go/channels_test.go:
```diff
@@ -15,7 +15,10 @@ import (
+	ch1ExtID     = sdk.Channel{ID: "fe6b4e92-cc98-425e-b0aa-000000000001", Name: "test1"}
+	ch2ExtID     = sdk.Channel{ID: "fe6b4e92-cc98-425e-b0aa-000000000002", Name: "test2"}
+	channel1     = sdk.Channel{Name: "test1"}
+	Prefix       = "fe6b4e92-cc98-425e-b0aa-"
```
Do you need to export this? `chPrefix` would be better in any case

aspnair replied:
moving to chPrefix

manuio replied:
btw, you could also use this prefix here:
```
 	ch1ExtID     = sdk.Channel{ID: chPrefix +-"000000000001", Name: "test1"} 
 ```

manuio commented on pkg/sdk/go/channels_test.go:
```diff
@@ -15,7 +15,10 @@ import (
-	channel      = sdk.Channel{ID: "001", Name: "test"}
+	ch1ExtID     = sdk.Channel{ID: "fe6b4e92-cc98-425e-b0aa-000000000001", Name: "test1"}
+	ch2ExtID     = sdk.Channel{ID: "fe6b4e92-cc98-425e-b0aa-000000000002", Name: "test2"}
+	channel1     = sdk.Channel{Name: "test1"}
```
I'd use `ch1` to be consistent with other namings

aspnair replied:
I am moving them to ch1, ch2, ch3 now.

manuio commented on pkg/sdk/go/things_test.go:
```diff
@@ -35,7 +35,8 @@ const (
 	metadata   = map[string]interface{}{"meta": "data"}
 	metadata2  = map[string]interface{}{"meta": "data2"}
-	thing      = sdk.Thing{ID: "001", Name: "test_device", Metadata: metadata}
+	thing1     = sdk.Thing{ID: "fe6b4e92-cc98-425e-b0aa-000000000001", Name: "test1", Metadata: metadata}
```
Same here, probably `th1` would be a more consistent name

aspnair replied:
Like above, moving them to th1, th2 for now.

manuio commented on pkg/sdk/go/things_test.go:
```diff
@@ -78,28 +79,28 @@ func TestCreateThing(t *testing.T) {
 			token:    token,
 			err:      nil,
-			location: "001",
+			location: "fe6b4e92-cc98-425e-b0aa-000000000001",
```
Use `ths1.ID` here

aspnair replied:
Moving to th1.ID

manuio commented on pkg/sdk/go/things_test.go:
```diff
@@ -78,28 +79,28 @@ func TestCreateThing(t *testing.T) {
 			token:    token,
 			err:      nil,
-			location: "002",
+			location: "123e4567-e89b-12d3-a456-000000000002",
```
Same here, use `th2.ID`

aspnair replied:
This is basically coming from pkg/uuid/mock.go.  So unlike above we can not use th2.ID, the test case would fail, if the value is changed.

manuio replied:
This could be done by using the Prefix const defined in the mock: `uuidMock.Prefix + "000000000002"` or as it's done in `pkg/uuid/mock` `fmt.Sprintf("%s%012d", uuidMock.Prefix, 2)`. But I think that the first one is more readable.

aspnair replied:
Updated with location: fmt.Sprintf("%s%012d", uuid.Prefix, 2),

manuio commented on things/api/things/http/endpoint_test.go:
```diff
@@ -126,7 +127,7 @@ func TestCreateThing(t *testing.T) {
 			auth:        token,
 			status:      http.StatusCreated,
-			location:    "/things/001",
+			location:    "/things/123e4567-e89b-12d3-a456-000000000001",
```
From where this location comes? Please, try to use consts as much as posible to make it understandable.

aspnair replied:
This location is also finally coming from pkg/uuid/mock.go only.  This Mock UUID is assigned during Thing / Channel creation.  More over the the same ID is used for storing inside things/mock directory also.  

We can see similar location used in twins/api/http/endpoint_twins_test.go:126 etc.

manuio replied:
Same here, simply use the const from the package. It's self explanatory.

manuio commented on things/api/things/http/endpoint_test.go:
```diff
@@ -142,7 +143,7 @@ func TestCreateThing(t *testing.T) {
 			auth:        token,
 			status:      http.StatusCreated,
-			location:    "/things/002",
+			location:    "/things/123e4567-e89b-12d3-a456-000000000003",
```
Same here

aspnair replied:
same as above

aspnair replied:
Done. Same as above

manuio commented on things/api/things/http/endpoint_test.go:
```diff
@@ -1296,7 +1304,7 @@ func TestCreateChannel(t *testing.T) {
 			auth:        token,
 			status:      http.StatusCreated,
-			location:    "/channels/001",
+			location:    "/channels/123e4567-e89b-12d3-a456-000000000001",
```
And here

aspnair replied:
same as above

aspnair replied:
Done.  Same as above

manuio commented on things/api/things/http/endpoint_test.go:
```diff
@@ -1328,7 +1336,7 @@ func TestCreateChannel(t *testing.T) {
 			auth:        token,
 			status:      http.StatusCreated,
-			location:    "/channels/002",
+			location:    "/channels/123e4567-e89b-12d3-a456-000000000002",
```
And here

aspnair replied:
same as above

aspnair replied:
Done. Same as above

manuio commented on things/api/things/http/endpoint_test.go:
```diff
@@ -35,6 +35,7 @@ const (
 	nameKey     = "name"
 	ascKey      = "asc"
 	descKey     = "desc"
+	Prefix      = "fe6b4e92-cc98-425e-b0aa-"
```
Why this const is exported? Can we maybe use `uuidPrefix`?

aspnair replied:
Ideally it should work when we use uuid.Prefix. This internal uuid.Prefix is used when we do not supply any external UUID. 
I was preferring to give an emphasis to show that we are using an external UUID.
Let me know your thought.  I am currently leaving it as it is and resolving it.

manuio replied:
There is no reason to export this constant if you are only using it in this package. Call it `prefix` or `uuidPrefix` please.

aspnair replied:
Now I realised your point :) There is no need to export Prefix.  Corrected and pushed, here and in things/service_test.go

manuio commented on things/api/things/http/endpoint_test.go:
```diff
@@ -881,7 +885,8 @@ func TestSearchThings(t *testing.T) {
 	for i := 0; i < 100; i++ {
 		name := "name_" + fmt.Sprintf("%03d", i+1)
-		ths, err := svc.CreateThings(context.Background(), token, things.Thing{Name: name, Metadata: map[string]interface{}{"test": name}})
+    id := fmt.Sprintf("%s%012d", Prefix, i + 1)
```
Align the tabulation

aspnair replied:
corrected it my mistake..

manuio commented on things/api/things/http/endpoint_test.go:
```diff
@@ -1048,7 +1053,10 @@ func TestListThingsByChannel(t *testing.T) {
 	data := []thingRes{}
 	for i := 0; i < 101; i++ {
-		ths, err := svc.CreateThings(context.Background(), token, thing)
+    id := fmt.Sprintf("%s%012d", Prefix, i + 1)
```
Align the tabulation

aspnair replied:
done

manuio commented on things/api/things/http/endpoint_test.go:
```diff
@@ -1899,7 +1907,10 @@ func TestListChannelsByThing(t *testing.T) {
 	channels := []channelRes{}
 	for i := 0; i < 101; i++ {
-		chs, err := svc.CreateChannels(context.Background(), token, channel)
+    id := fmt.Sprintf("%s%012d", Prefix, i + 1)
```
Align the tabulation

aspnair replied:
updated

manuio commented on things/service_test.go:
```diff
@@ -24,11 +24,14 @@ const (
 	token      = "token"
 	token2     = "token2"
 	n          = uint64(10)
+	Prefix     = "fe6b4e92-cc98-425e-b0aa-"
```
Same here, why this is exported? Maybe we can use `uuidPrefix`

aspnair replied:
same as above.

manuio commented on things/redis/streams_test.go:
```diff
@@ -36,6 +36,7 @@ const (
 	channelCreate = channelPrefix + "create"
 	channelUpdate = channelPrefix + "update"
 	channelRemove = channelPrefix + "remove"
+	Prefix        = "fe6b4e92-cc98-425e-b0aa-"
```
Is this used? Does it have to be exported?

aspnair replied:
Correct.. I saw that it is not getting used.  just removed it.

manuio commented on things/service_test.go:
```diff
@@ -24,11 +24,14 @@ const (
-	channel = things.Channel{Name: "test"}
+	thingList  = [n]things.Thing{}
+	channel    = things.Channel{Name: "test"}
+	thsExtID   = []things.Thing{{ID: "fe6b4e92-cc98-425e-b0aa-000000000001", Name: "a"}, {ID: "fe6b4e92-cc98-425e-b0aa-000000000002", Name: "b"}}
```
You could use the prefix here: `prefix + "000000000001"` and ``prefix + "000000000002"``

aspnair replied:
done.. resubmitted the patch also.

drasko commented on things/service.go:
```diff
@@ -167,7 +178,7 @@ func (ts *thingsService) CreateThings(ctx context.Context, token string, things
 
 	for i := range things {
-		things[i].ID, err = ts.idProvider.ID()
+		things[i].ID, err = ts.provideUUID(things[i].ID)
```
There is some code smell around this line... You are providing `things[i].ID` as a param, then checking in the function if it is empty string, and so on.

Why not checking right away if it is an empty string, and calling `ts.idProvider.ID()` only if it is not?

dborovcanin replied:
While there is nothing wrong with the code, please consider changing to something like:
```go
func (ts *thingsService) CreateThings(ctx context.Context, token string, things ...Thing) ([]Thing, error) {
	res, err := ts.auth.Identify(ctx, &mainflux.Token{Value: token})
	if err != nil {
		return []Thing{}, errors.Wrap(ErrUnauthorizedAccess, err)
	}
	email := res.GetEmail()

	for i, th := range things {
		if th.ID != "" {
			id, err := ts.idProvider.ID()
			if err != nil {
				return []Thing{}, errors.Wrap(ErrCreateUUID, err)
			}
			things[i].ID = id
		}

		if th.Key == "" {
			key, err := ts.idProvider.ID()
			if err != nil {
				return []Thing{}, errors.Wrap(ErrCreateUUID, err)
			}
			things[i].Key = key
		}

		things[i].Owner = email
	}

	return ts.things.Save(ctx, things...)
}
````
It looks a bit cleaner and better fits in the codebase.

aspnair replied:
Done

drasko commented on things/service.go:
```diff
@@ -253,7 +264,7 @@ func (ts *thingsService) CreateChannels(ctx context.Context, token string, chann
 
 	for i := range channels {
-		channels[i].ID, err = ts.idProvider.ID()
+		channels[i].ID, err = ts.provideUUID(channels[i].ID)
```
Same here

dborovcanin replied:
I agree with @drasko. You can use code similar to the snippet for `CreateThings`. It's repetitive, I agree, but also a bit cleaner and simpler. You can also completely delete `provideUUID` method from the service.

aspnair replied:
done

dborovcanin commented on things/service_test.go:
```diff
@@ -557,6 +578,18 @@ func TestCreateChannels(t *testing.T) {
+			err:    nil,
+		},
+		{
+			desc:   "create new channels with external wrong UUID",
```
create new channels with invalid external UUID

aspnair replied:
done

dborovcanin commented on things/mocks/commons.go:
```diff
@@ -83,3 +86,17 @@ func sortChannels(pm things.PageMetadata, chs []things.Channel) []things.Channel
+		slNum = ID[len(ID) - 6:]
+	} else {
+		slNum = ID
+	}
```
Please avoid using `else`.

aspnair replied:
done

manuio commented on things/api/things/http/requests.go:
```diff
@@ -21,14 +22,33 @@ type createThingReq struct {
+func validateUUID(extID string) (err error) {
+	if extID == "" {
+		return nil
+	}
```
Move this check from here. An empty string should return an error

aspnair replied:
But, if we remove this check, it is breaking the testcases.
Because if extID is null (in the case of when there is no extID supplied), it would goto uuid.FromString and return error.



go test -mod=vendor -v -race -tags test -coverprofile=profile.out -covermode=atomic github.com/mainflux/mainflux/things/api/things/http
=== RUN   TestCreateThing
    endpoint_test.go:211: 
                Error Trace:    endpoint_test.go:211
                Error:          Not equal: 
                                expected: 201
                                actual  : 400
                Test:           TestCreateThing
                Messages:       add valid thing: expected status code 201 got 400
    endpoint_test.go:212: 
                Error Trace:    endpoint_test.go:212
                Error:          Not equal: 
                                expected: "/things/123e4567-e89b-12d3-a456-000000000001"
                                actual  : ""
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -/things/123e4567-e89b-12d3-a456-000000000001
                                +
                Test:           TestCreateThing
                Messages:       add valid thing: expected location /things/123e4567-e89b-12d3-a456-000000000001 got 
    endpoint_test.go:211: 
                Error Trace:    endpoint_test.go:211
                Error:          Not equal: 
                                expected: 409
                                actual  : 400
                Test:           TestCreateThing
                Messages:       add thing with existing key: expected status code 409 got 400
    endpoint_test.go:211: 
                Error Trace:    endpoint_test.go:211
                Error:          Not equal: 
                                expected: 201
                                actual  : 400
                Test:           TestCreateThing
                Messages:       add thing with empty JSON request: expected status code 201 got 400
    endpoint_test.go:212: 
                Error Trace:    endpoint_test.go:212
                Error:          Not equal: 
                                expected: "/things/123e4567-e89b-12d3-a456-000000000003"
                                actual  : ""
                            
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -/things/123e4567-e89b-12d3-a456-000000000003
                                +
                Test:           TestCreateThing
                Messages:       add thing with empty JSON request: expected location /things/123e4567-e89b-12d3-a456-000000000003 got 
    endpoint_test.go:211: 
                Error Trace:    endpoint_test.go:211
                Error:          Not equal: 
                                expected: 401
                                actual  : 400
                Test:           TestCreateThing
                Messages:       add thing with invalid auth token: expected status code 401 got 400
    endpoint_test.go:211: 
                Error Trace:    endpoint_test.go:211
                Error:          Not equal: 
                                expected: 401
                                actual  : 400
                Test:           TestCreateThing
                Messages:       add thing with empty auth token: expected status code 401 got 400
--- FAIL: TestCreateThing (0.03s)
=== RUN   TestCreateThings
--- PASS: TestCreateThings (0.03s)
=== RUN   TestUpdateThing
--- PASS: TestUpdateThing (0.03s)
=== RUN   TestUpdateKey
--- PASS: TestUpdateKey (0.03s)
=== RUN   TestViewThing
--- PASS: TestViewThing (0.02s)
=== RUN   TestListThings
--- PASS: TestListThings (0.06s)
=== RUN   TestSearchThings
--- PASS: TestSearchThings (0.05s)
=== RUN   TestListThingsByChannel
--- PASS: TestListThingsByChannel (1.06s)
=== RUN   TestRemoveThing
--- PASS: TestRemoveThing (0.01s)
=== RUN   TestCreateChannel
--- PASS: TestCreateChannel (0.02s)
=== RUN   TestCreateChannels
--- PASS: TestCreateChannels (0.05s)
=== RUN   TestUpdateChannel
--- PASS: TestUpdateChannel (0.03s)
=== RUN   TestViewChannel
--- PASS: TestViewChannel (0.01s)
=== RUN   TestListChannels
--- PASS: TestListChannels (0.09s)
=== RUN   TestListChannelsByThing
--- PASS: TestListChannelsByThing (0.06s)
=== RUN   TestRemoveChannel
--- PASS: TestRemoveChannel (0.01s)
=== RUN   TestConnect
--- PASS: TestConnect (0.02s)
=== RUN   TestCreateConnections
--- PASS: TestCreateConnections (0.05s)
=== RUN   TestDisconnectList
--- PASS: TestDisconnectList (0.05s)
=== RUN   TestDisconnnect
--- PASS: TestDisconnnect (0.03s)
FAIL
coverage: 85.6% of statements
FAIL    github.com/mainflux/mainflux/things/api/things/http     1.804s
FAIL

aspnair replied:
IIRC as it was done some time before - validateUUID() should be successful if we have an empty UUID or a valid UUID.

manuio commented on things/api/things/http/requests.go:
```diff
@@ -21,14 +22,33 @@ type createThingReq struct {
 		return things.ErrUnauthorizedAccess
 	}
 
+	if validateUUID(req.ID) != nil {
```
`if req.ID != ""  && validateUUID(req.ID) != nil {`

aspnair replied:
done

manuio replied:
Looks like you did it in a wrong line. As I said before revert line 44 and add the check for empty ID here.

manuio commented on things/api/things/http/requests.go:
```diff
@@ -51,6 +71,10 @@ func (req createThingsReq) validate() error {
 	}
 
 	for _, thing := range req.Things {
+		if validateUUID(thing.ID) != nil {
```
`if req.ID != ""  && validateUUID(thing.ID) != nil {`

aspnair replied:
done

aspnair replied:
if thing.ID != "" && validateUUID(thing.ID) != nil {

manuio commented on things/api/things/http/requests.go:
```diff
@@ -111,6 +136,10 @@ func (req createChannelReq) validate() error {
 		return things.ErrUnauthorizedAccess
 	}
 
+	if validateUUID(req.ID) != nil {
```
`if req.ID != ""  && validateUUID(req.ID) != nil {`

aspnair replied:
done

manuio commented on things/api/things/http/requests.go:
```diff
@@ -133,6 +162,10 @@ func (req createChannelsReq) validate() error {
 	}
 
 	for _, channel := range req.Channels {
+		if validateUUID(channel.ID) != nil {
```
`if req.ID != ""  && validateUUID(channel.ID) != nil {`

aspnair replied:
if channel.ID != "" && validateUUID(channel.ID) != nil {
done

manuio commented on things/api/things/http/requests.go:
```diff
@@ -21,14 +22,33 @@ type createThingReq struct {
+
 func (req createThingReq) validate() error {
-	if req.token == "" {
+	if req.ID != "" && validateUUID(req.ID) != nil {
```
Why did you removed the token check? Revert this and move `req.ID != "" && validateUUID(req.ID) != nil ` to line 48 please.

aspnair replied:
My mistake.  Somehow CI did not catch that too :)  Corrected it now..

manuio reviewed (dismissed):
LGTM

drasko commented on pkg/sdk/go/things_test.go:
```diff
@@ -127,8 +126,16 @@ func TestCreateThings(t *testing.T) {
+		th1,
+		th2,
+	}
+	thExtID := []sdk.Thing {
```
Call these `thsExtID`, so that we know that this is an array of several things.

aspnair replied:
done

drasko commented on pkg/sdk/go/things_test.go:
```diff
@@ -127,8 +126,16 @@ func TestCreateThings(t *testing.T) {
+		sdk.Thing{ID: th1.ID, Name: "1", Key: "1", Metadata:metadata},
+		sdk.Thing{ID: th2.ID, Name: "2", Key: "2", Metadata:metadata},
+	}
+	thWrongExtID := []sdk.Thing {
```
Same here

aspnair replied:
done

aspnair committed:
MF-1425 Enhancement for external UUID

MF-1425 Enhancement to supply an external UUID for Things and Channels.
- New local function validateUUID() to use gofrs/uuid FromString()
  to validate the supplied UUID, in Things HTTP Requests
- New local function provideUUID() to generate a random UUID or to take
  an external supplied UUID, in Things Service layer.
- If the tag "id" is not used, it uses the current default method
  of generating a random UUID using the configured provider.

This "id" tag could be used for four different endpoints, namely,
- things
- things/bulk
- channels
- channels/bulk

Usage:
$ uuidgen
99336f21-bd45-4e47-87ab-f57505bf5ef2

$ curl -s -S -i -X POST -H "Content-Type: application/json" \
  -H "Authorization: ${TOKEN}" http://localhost/things \
  -d '{"id": "99336f21-bd45-4e47-87ab-f57505bf5ef2", "name": "t1"}'

$ curl -s -S -i -X GET -H "Content-Type: application/json" \
  -H "Authorization: ${TOKEN}" http://localhost/things
{"total":1,"offset":0,"limit":10,"order":"",
"direction":"","things":[{"id":"99336f21-bd45-4e47-87ab-f57505bf5ef2",
"name":"t1","key":"a87563c2-7a35-449d-b50a-c7a261a61aa5"}]}

Signed-off-by: Anand Sivaram Palassery <aspnair@gmail.com>

aspnair committed:
MF-1425 - Test cases changes for SDK

MF-1425 Enhancement to supply an external UUID for Things and Channels.
These are the new testcases added for
- Things Service Testcases
- SDK Things and Channel Testcases

Signed-off-by: Anand Sivaram Palassery <aspnair@gmail.com>

aspnair committed:
MF-1425 - Fixing Testcases

MF-1425 Enhancement to supply an external UUID for Things and Channels.
Because of the previous commits, the testcases were getting failed
because the way ID was modified.  This change is to make sure that
all testcases are revisited to get them fixed.

Signed-off-by: Anand Sivaram Palassery <aspnair@gmail.com>

aspnair committed:
MF-1425 - Fixing review comments

Fixing the review comments provided.

Signed-off-by: Anand Sivaram Palassery <aspnair@gmail.com>

buraksekili commented on things/service.go:
```diff
@@ -176,21 +165,26 @@ func (ts *thingsService) CreateThings(ctx context.Context, token string, things
+	email := res.GetEmail()
 
-	for i := range things {
-		things[i].ID, err = ts.provideUUID(things[i].ID)
```
Why did you revert to `ts.idProvider.ID()` instead of `provideUUID`? Sorry, if I missed any previous discussions about it. I am just asking out of curiosity.

aspnair replied:
Hello @buraksekili , I removed provideUUID as per the comment given by @manuio and @dusanb94

aspnair committed:
MF-1425 - Fixing more review comments

Signed-off-by: Anand Sivaram Palassery <aspnair@gmail.com>

manuio commented on api/things.yml:
```diff
@@ -610,7 +610,10 @@ components:
           type: string
           format: uuid
-          description: Thing unique identifier
+          description: Thing unique identifier.  This can be either
```
This doesn't look correct, you modified Identity schema. Please add the ID in `ThingReqSchema`:
```
    ThingReqSchema:
      type: object
      properties:
        id:
          $ref: "#/components/schemas/ID"
        key:
          $ref: "#/components/schemas/Key"
        name:
          type: string
          description: Free-form thing name.
        metadata:
          type: object
          description: Arbitrary, object-encoded thing's data.
```

And  add the schema:
```
  schemas:
    ID:
      type: string
      format: uuid
      description: |
        Thing ID that is used for thing identity. If there is
        not one provided service will generate one in UUID
        format.
```

manuio committed:
Merge branch 'master' into master
